crates/ra_hir_ty/src/traits.rs

```diff
@@ -64,11 +64,15 @@ struct ChalkContext<'a, DB> {
     krate: CrateId,
 }
 
-pub(crate) fn trait_solver_query(
-    db: &(impl HirDatabase + salsa::Database),
-    krate: CrateId,
-) -> TraitSolver {
-    db.salsa_runtime().report_untracked_read();
+pub(crate) fn trait_solver_query(db: &impl HirDatabase, krate: CrateId) -> TraitSolver {
+    // Chalk solver uses some internal caching, make sure to invalidate it when,
+    // roughly, a set of top-level items changes. We probably want some more
+    // precise and more correct invalidation strategy here!
+    let _ = db.crate_def_map(krate);
```

Comment:
I don't think crate def map is enough -- the solver needs to access things like where clauses on items and will cache clauses based on that

Reply:
Ahhh, right  :man_facepalming: =(